internal/pkg/service/common/etcdop/op/atomic_ctx.go

```diff
@@ -0,0 +1,74 @@
+package op
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+const (
+	actualAtomicOpCtxKey  = ctxKey("actualAtomicOp")
+	atomicOpMaxReadLevels = 10
+)
+
+type ctxKey string
+
+type atomicOpCore = AtomicOpCore
+
+// ActualAtomicOp - aux struct, part of the context, to provide actual atomic operation to be extended.
+type ActualAtomicOp struct {
+	*atomicOpCore
+	*atomicOpStore
+	closed bool
+}
+
+type atomicOpStore struct {
+	store map[string]any
+}
+
+// AtomicOpFromCtx gets actual atomic operation from the context.
+//
+// It can be used to add some additional READ operations based on result from a previous READ operation via *AtomicOpCore methods.
+// See AtomicOp.Do method and TestAtomicFromCtx_Complex for details.
+//
+// In addition, it is possible to set auxiliary key/value pairs that should be available in the atomic operation.
+// See atomicOpStore struct.
+func AtomicOpFromCtx(ctx context.Context) *ActualAtomicOp {
+	actualOp, ok := ctx.Value(actualAtomicOpCtxKey).(*ActualAtomicOp)
+	if !ok {
+		panic(errors.New("no atomic operation found in the context"))
+	}
+	if actualOp.closed {
+		panic(errors.New("atomic operation in the context is closed"))
+	}
+	return actualOp
+}
+
+func newActualAtomicOp(core *AtomicOpCore, store *atomicOpStore) *ActualAtomicOp {
```

Comment:
What does Actual represent? The newAtomicOp is not actual but fake/old?

Reply:
It is "something" that represents in progress atomic operation. It can be obtained from the context, plus it implements the `store` methods.

https://github.com/keboola/keboola-as-code/blob/746f67df64188dc4de776baa4858614f378da6c1/internal/pkg/service/common/etcdop/op/atomic_do.go#L52-L90

----------------------

Suggestions for better naming and comments are welcome. cc @jachym-tousek-keboola @hosekpeter

Reply:
:arrow_right:  `AtomicOpContext`